source/server/listener_impl.cc

```diff
@@ -26,19 +26,36 @@
 namespace Envoy {
 namespace Server {
 
-ListenSocketFactoryImplBase::ListenSocketFactoryImplBase(
-    ListenerComponentFactory& factory, Network::Address::InstanceConstSharedPtr local_address,
-    Network::Address::SocketType socket_type, const Network::Socket::OptionsSharedPtr& options,
-    bool bind_to_port, const std::string& listener_name)
-    : factory_(factory), local_address_(local_address), socket_type_(socket_type),
-      options_(options), bind_to_port_(bind_to_port), listener_name_(listener_name) {}
-
-Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApplyOptions() {
+ListenSocketFactoryImpl::ListenSocketFactoryImpl(ListenerComponentFactory& factory,
+                                                 Network::Address::InstanceConstSharedPtr address,
+                                                 Network::Address::SocketType socket_type,
+                                                 const Network::Socket::OptionsSharedPtr& options,
+                                                 bool bind_to_port,
+                                                 const std::string& listener_name, bool reuse_port)
+    : factory_(factory), local_address_(address), socket_type_(socket_type), options_(options),
+      bind_to_port_(bind_to_port), listener_name_(listener_name), reuse_port_(reuse_port) {
+
+  // If reuse_port_ is false, create a socket here which will be used by all worker
+  // threads; otherwise, if port is 0, still need to create a socket here, because
+  // a random port will be assigned, then all worker threads should use same port.
+  if (!reuse_port_ || local_address_->ip()->port() == 0) {
```

Comment:
nit: ASSERT(socket_type == Stream)

Reply:
socket_type could be Datagram here.

Reply:
I think Datagram socket will for sure reuse port, no? My point is enforcing this assumption here.

Reply:
Is it cleaner to have two implementations one for providing reuse port socket(UDP or TCP), the other for shared socket(only TCP)? I could foresee that these two implementations will be totally different, so there is no point of mixing them together.

Reply:
reuse_port is just an option, I don't think it's necessary to have two factory implementations because of it.

Reply:
> nit: ASSERT(socket_type == Stream)

added like:
  if (socket_type_ == Network::Address::SocketType::Datagram) {
    ASSERT(reuse_port_ == true);
  }

Reply:
I think we need to figure out how to create socket for reuse port first before looking into these implementation details.